mne/io/neuralynx/neuralynx.py

```diff
@@ -98,31 +139,133 @@ def __init__(
+            gap_indicator = np.concatenate(
+                [
+                    np.full(len(start_times), fill_value=0),
+                    np.full(len(gap_starts), fill_value=1),
+                ]
+            )
+            gap_indicator = gap_indicator[all_starts_ids].astype(bool)
+
+            # store this in a dict to be passed to _raw_extras
+            seg_gap_dict = {
+                "gap_n_samps": gap_n_samps,
+                "isgap": gap_indicator,  # 0 (data segment) or 1 (gap segment)
+            }
+
+        else:
+            logger.info(
+                f"All Neo segments temporally continuous at {delta} sec precision."
+            )
+
+        valid_segment_sizes = [
+            nlx_reader.get_signal_size(block_id, i) for i in range(n_segments)
+        ]
+
+        if has_gaps:
+            sizes_sorted = np.concatenate([valid_segment_sizes, gap_n_samps])[
+                all_starts_ids
             ]
+        else:
+            sizes_sorted = np.array(valid_segment_sizes)
+
+        # now construct an (n_samples,) indicator variable
+        sample2segment = np.concatenate(
+            [np.full(shape=(n,), fill_value=i) for i, n in enumerate(sizes_sorted)]
         )
 
+        # construct Annotations()
+        if has_gaps:
+            gap_seg_ids = np.unique(sample2segment)[gap_indicator]
+            gap_start_ids = np.array(
+                [np.where(sample2segment == seg_id)[0][0] for seg_id in gap_seg_ids]
+            )
+
+            # recreate time axis for gap annotations
+            mne_times = np.arange(0, len(sample2segment)) / info["sfreq"]
+
+            assert len(gap_start_ids) == len(gap_n_samps)
+            annotations = Annotations(
+                onset=[mne_times[onset_id] for onset_id in gap_start_ids],
+                duration=[
+                    mne_times[onset_id + (n - 1)] - mne_times[onset_id]
+                    for onset_id, n in zip(gap_start_ids, gap_n_samps)
+                ],
+                description=["BAD_ACQ_SKIP" for _ in gap_start_ids],
+            )
+        else:
+            annotations = None
+
         super(RawNeuralynx, self).__init__(
             info=info,
-            last_samps=[n_total_samples - 1],
+            last_samps=[sizes_sorted.sum() - 1],
             filenames=[fname],
             preload=preload,
-            raw_extras=[dict(smp2seg=sample2segment, exclude_fnames=exclude_fnames)],
+            raw_extras=[
+                dict(
+                    smp2seg=sample2segment,
+                    exclude_fnames=exclude_fnames,
+                    segment_sizes=sizes_sorted,
+                    seg_gap_dict=seg_gap_dict,
+                )
+            ],
         )
 
+        self.set_annotations(annotations)
+
     def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
         """Read a chunk of raw data."""
+        from neo import Segment
         from neo.io import NeuralynxIO
+        from quantities import Hz
```

Comment:
I had to look this up to make sure, so let's save the next dev the trouble
```suggestion
        # quantities is a dependency of neo so we are guaranteed it exists
        from quantities import Hz
```